### PR TITLE
Bug 1908342: Add pv_collector_total_pv_count storage metric

### PR DESCRIPTION
**Pull request by tsmetana**
This is a metric that allows us to get counts fo the volume plugins and volume types being used.

* [ ] I added CHANGELOG entry for this change.
* [x] No user facing changes, so no entry in CHANGELOG was needed.

### CODE REVIEW

**Comment by lilic** [comment]
@tsmetana thanks for the PR, a couple of questions:
1. did you follow the telemetry doc and get approval from folks like Clayton before opening the PR?
2. Where does this metric come from? `pv_collector_total_pv_count`? 
3. Nit pick on best practices: Note it should be pv_collector_total, no need for count at the end only https://prometheus.io/docs/practices/naming/ so `pv_collector_total`.

**Comment by tsmetana** [comment]
I think I missed something... I followed the README that says
> To add new metrics to be sent via telemetry, simply add a selector that matches the time-series to be sent in
> manifests/0000_50_cluster-monitoring-operator_04-config.yaml.
So I'll ask Clayton, I guess with teh PR it's at least more obvious what I want.

The metric has been added by [kubernetes PR#95719](https://github.com/kubernetes/kubernetes/pull/95719) (so we'll get it in 4.7 with rebase to 1.20) which followed the naming convention used by the other PV controller metrics: https://github.com/kubernetes/kubernetes/pull/95719/files#diff-3b653bbcafd5e626334010f66e6754eeab8dc1032ccc487276e80e965d7632e9R35-R38

Thanks!

**Comment by smarterclayton** [comment]
How many series per cluster?

I would prefer this be a recording rule that normalizes and trims excess labels so we know the cardinality is bounded.  Also, the name should be consistent with `cluster:kube_persistentvolumeclaim_resource_requests_storage_bytes:provisioner:sum` - with  `cluster:kube_persistentvolume_....` as the prefix

**Comment by tsmetana** [comment]
I hope I've managed to address the feedback with the new changes.

**Comment by tsmetana** [comment]
@smarterclayton PTAL...

**Comment by lilic** [comment]
> The metric has been added by kubernetes PR#95719 (so we'll get it in 4.7 with rebase to 1.20) which followed the naming convention used by the other PV controller metrics: https://github.com/kubernetes/kubernetes/pull/95719/files#diff-3b653bbcafd5e626334010f66e6754eeab8dc1032ccc487276e80e965d7632e9R35-R38

Looks like no one from sig instrumentation actually approved that PR 🤔

**Comment by tsmetana** [comment]
> Looks like no one from sig instrumentation actually approved that PR

True. It was in the merge queue for some 20 days though. Actually https://github.com/kubernetes/kubernetes/pull/59646 that added the other pv_collector metrics was also reviewed by the storage people only. It's quite old, I'm not sure if sig-instrumentation existed back then. And yes, it explains the metric names...

**Comment by smarterclayton** [comment]
Example from e2e run:

```

{plugin_name="kubernetes.io/local-volume",volume_mode="Filesystem"} | 8
-- | --
{plugin_name="kubernetes.io/aws-ebs",volume_mode="Filesystem"} | 5
{plugin_name="kubernetes.io/nfs",volume_mode="Filesystem"} | 4
{plugin_name="kubernetes.io/csi:csi-hostpath-e2e-provisioning-5234",volume_mode="Filesystem"} | 2
{plugin_name="kubernetes.io/glusterfs",volume_mode="Filesystem"} | 2
{plugin_name="kubernetes.io/csi:csi-hostpath-e2e-provisioning-5911",volume_mode="Block"} | 2
{plugin_name="kubernetes.io/csi:csi-hostpath-e2e-snapshotting-8131",volume_mode="Filesystem"} | 2
{plugin_name="kubernetes.io/csi:csi-hostpath-e2e-snapshotting-8070",volume_mode="Filesystem"} | 2
{plugin_name="kubernetes.io/csi:csi-hostpath-e2e-provisioning-7263",volume_mode="Filesystem"} | 2
{plugin_name="kubernetes.io/csi:csi-hostpath-e2e-provisioning-8342",volume_mode="Block"} | 2
... // about 30 others from e2e tests
```

Cardinality will remain low on normal systems.

/approve
/lgtm

/retest

**Comment by smarterclayton** [comment]
/retest

Please link to a bug

**Comment by dgrisonnet** [comment]
@tsmetana the CI error is valid, you need to run `make generate` and push the modifications to the `assets/prometheus-k8s/rules.yaml` file.

**Comment by tsmetana** [comment]
> @tsmetana the CI error is valid, you need to run `make generate` and push the modifications to the `assets/prometheus-k8s/rules.yaml` file.

Thanks. I didn't realize I had to rebase before doing so... Hope it's finally OK.

**Comment by tsmetana** [comment]
@smarterclayton @dgrisonnet @lilic thanks for help. I guess I need to re-lgtm this again to get it finally merged...

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,7 @@
 - [#970](https://github.com/openshift/cluster-monitoring-operator/pull/970) Bump prometheus-operator v0.43.0.
 - [#971](https://github.com/openshift/cluster-monitoring-operator/pull/971) Enable `hwmon` in node-exporter for hardware sensor data collection
 - [#983](https://github.com/openshift/cluster-monitoring-operator/pull/983) Remove deprecated user workload configuration
+- [#994](https://github.com/openshift/cluster-monitoring-operator/pull/994) Collect PV volume plugin type and volume mode metrics
```
Comment (lilic) [inline CHANGELOG.md:14]: Don't think this needs to be included in the release notes. We use the CHANGELOG to help our docs team to inform things users need to know. This is more for our internal devs so lets remove this.

**Review by lilic — COMMENTED** [review]
Did not see that before.

Afterwards lgtm

File: CHANGELOG.md (hunk 1 above)
Comment (tsmetana) [reply CHANGELOG.md]: Removed

**Review by lilic — COMMENTED** [review]
/lgtm

**Pull request merged**